javascript/ql/src/semmle/javascript/Expr.qll

```diff
@@ -161,6 +161,39 @@ class Expr extends @expr, ExprOrStmt, ExprOrType, AST::ValueNode {
    * file that was extracted without type information.
    */
   Type getType() { ast_node_type(this, result) }
+
+  /**
+   * Holds if the syntactic context that the expression appears in relies on the expression
+   * being non-null/non-undefined.
+   */
+  predicate inNullSensitiveContext() {
+    exists(ExprOrStmt ctx |
+      this = ctx.(PropAccess).getBase()
+      or
+      this = ctx.(IndexExpr).getPropertyNameExpr()
+      or
+      this = ctx.(InvokeExpr).getCallee()
+      or
+      this = ctx.(BinaryExpr).getAnOperand() and
+      not ctx instanceof LogicalBinaryExpr and
+      not ctx instanceof EqualityTest
+      or
+      this = ctx.(UnaryExpr).getOperand() and
+      not ctx instanceof LogNotExpr
+      or
+      this = ctx.(UpdateExpr).getOperand()
+      or
+      this = ctx.(CompoundAssignExpr).getLhs()
+      or
+      this = ctx.(CompoundAssignExpr).getRhs()
+      or
+      this = ctx.(AssignExpr).getRhs()
```

Comment:
This seems a little odd; why is the right-hand side of an assignment null-sensitive?

Reply:
Do you perhaps want the left-hand side to be a destructuring pattern?

Reply:
Blergh, yes, copying error. Thank you for catching this. :)